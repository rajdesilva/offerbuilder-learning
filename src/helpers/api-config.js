import { init as initApm } from "@elastic/apm-rum";

export const baseURL = {
  API_URL_BASE: process.env.REACT_APP_BASE_API_URL,
  LOGIN_URL: process.env.REACT_APP_LOGIN_URL,
  LOGOUT_URL: process.env.REACT_APP_LOGOUT_URL,
  PROFILE_URL: process.env.REACT_APP_USER_PROFILE_URL,
  HISTORY_URL: process.env.REACT_APP_USER_HISTORY_URL,
};

export const apm = initApm({
  // Set required service name (allowed characters: a-z, A-Z, 0-9, -, _, and space)
  serviceName: process.env.REACT_APP_APM_SERVICE_NAME,
  serverUrl: process.env.REACT_APP_APM_SERVICE_URL,

  // Set service version (required for sourcemap feature)
  serviceVersion: process.env.REACT_APP_APM_SERVICE_VERSION,
});