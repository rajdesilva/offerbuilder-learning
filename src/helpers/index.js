export * from './api-config';
export * from './history';
