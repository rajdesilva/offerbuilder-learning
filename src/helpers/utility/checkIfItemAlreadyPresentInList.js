export const checkIfItemAlreadyPresentInList = (
  itemToCheck,
  itemList
) => {
  if(itemList && itemToCheck && itemToCheck.id) {
    return (
      itemList.findIndex(
        (item) => item.id === itemToCheck.id
      ) !== -1
    );
  }
  return false;
};
