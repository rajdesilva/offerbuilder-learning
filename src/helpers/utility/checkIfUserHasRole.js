import { store } from "../../redux/store";

export const checkIfUserHasRole = (roleToFind) => {
  try {
    const userRoles =
      window.getValue(store.getState(), "userinfo.userDetails.roles") || [];
    return userRoles.findIndex((role) => role === roleToFind) !== -1;
  } catch (error) {
    console.log("error = ", error.toString());
  }
  return false;
};
