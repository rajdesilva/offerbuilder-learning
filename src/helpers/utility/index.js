export { createSearchQueryURL } from './createSearchQueryURL';
export { isParentOrChildrenPropertyAdded } from './isParentOrChildrenPropertyAdded';
export { isPropertySelected } from './isPropertySelected';
export { mapBrands } from './mapBrands';
export { mapOnlySuppliers } from './mapOnlySuppliers';
export { parseFormDescriptionsAndImagesForOffer } from './parseFormDescriptionsAndImagesForOffer';
export { removePropertyForSelectedChannels } from './removePropertyForSelectedChannels';
export { removePropertyForSelectedSuppliers } from './removePropertyForSelectedSuppliers';
export { getParsedProperty } from './getParsedProperty';
export { getFormattedValue } from './getFormattedValue';
export { getValue } from './getValue';
export { parsePropertyMarketingInfoToFormData } from './parsePropertyMarketingInfoToFormData';
export { createOfferRequestObject } from './createOfferRequestObject';
export { getParsedOffers } from './getParsedOffers';
export { getListOfPropertiesForSupplier } from './getListOfPropertiesForSupplier';
export { getListOfPropertiesForChannels } from './getListOfPropertiesForChannels';
export { getInitializedState } from './getInitializedState';
export { getParsedNewOfferSettingsObject } from './getParsedNewOfferSettingsObject';
export { getListOfPropertiesForStorefront } from './getListOfPropertiesForStorefront';
export { checkIfMultipleChannelPresentForTarget } from './checkIfMultipleChannelPresentForTarget';
export { checkIfMultipleSupplierPresentForTarget } from './checkIfMultipleSupplierPresentForTarget';
export { checkIfMultipleStorefrontPresentForTarget } from './checkIfMultipleStorefrontPresentForTarget';
export { removePropertyForSelectedStorefront } from './removePropertyForSelectedStorefront';
export { checkIfMultipleBrandPresentForTarget } from './checkIfMultipleBrandPresentForTarget';
export { getListOfPropertiesForBrand } from './getListOfPropertiesForBrand';
export { getSearchStateFromRedux } from './getSearchStateFromRedux';
export { getObjectStringify } from './getObjectStringify';
export { parseOfferMarketingFromOffer } from './parseOfferMarketingFromOffer';
export { parseOfferMarketingLanguagesFromOffer } from './parseOfferMarketingLanguagesFromOffer';
export { getFilterData } from './getFilterData';
export { getSearchAndFilterQueryURL } from './getSearchAndFilterQueryURL';
export { getSearchQueryURLForEditAddProperties } from "./getSearchQueryURLForEditAddProperties";
export { isSavedOfferProperty } from './isSavedOfferProperty';
export { getParsedPropertiesForDeeplinkSettings } from "./getParsedPropertiesForDeeplinkSettings";
export { resetApplicationReduxAndStorage } from "./resetApplicationReduxAndStorage";
export { updateSettingsBrandWithSearchedValues } from "./updateSettingsBrandWithSearchedValues";
export { checkIfUserHasRole } from "./checkIfUserHasRole";
export { getUserOfferBuilderRole } from "./getUserOfferBuilderRole";
export { checkIfNewPropertiesAdded } from "./checkIfNewPropertiesAdded";
export { getAllBrandsWithoutKeysFromTempBrands } from "./getAllBrandsWithoutKeysFromTempBrands";
export { checkIfItemAlreadyPresentInList } from "./checkIfItemAlreadyPresentInList";
export { checkIfTravellingDateValidAsPerBookingDate } from "./checkIfTravellingDateValidAsPerBookingDate";
export { checkIfPropertiesHaveMultipleLocations } from "./checkIfPropertiesHaveMultipleLocations";
export { displayMargin } from "./displayMargin";
export { getFilterCount } from "./getFilterCount";
export { labelForBrands } from "./labelForBrands";
export { getDistanceFromLatLonInKm } from "./getDistanceFromLatLonInKm";
export { getTrustYouIdListFromProperties } from "./getTrustYouIdListFromProperties";
export { getSearchParamsToFetchMarketPrice } from "./getSearchParamsToFetchMarketPrice";
export { getMarketPriceForPropertyFromList } from "./getMarketPriceForPropertyFromList";
export { getImageFilterCount } from "./getImageFilterCount";
export { getSearchAndFilterImagesQueryUrl } from "./getSearchAndFilterImagesQueryUrl";
export { getImageName } from "./getImageName";
export { updateReduxWithOfferData } from "./updateReduxWithOfferData";
export { getOfferTypeFromPropertiesSelected } from "./getOfferTypeFromPropertiesSelected";