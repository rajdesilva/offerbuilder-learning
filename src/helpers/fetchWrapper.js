import { message } from "antd";
import { createIntl, createIntlCache } from "@formatjs/intl";
import * as languages from "../langs/";
import { baseURL } from "./api-config";
import { history } from "./history";

// This is optional but highly recommended
// since it prevents memory leak
const cache = createIntlCache();

// Create the `intl` object
const intl = createIntl(
  {
    // Locale of the application
    // TODO:
    locale: "en",
    // Locale of the fallback defaultMessage
    defaultLocale: "en",
    messages: languages["en"],
  },
  cache
);
const defaultRequest = {
  credentials: "include",
  headers: {
    "Content-Type": "application/json",
  },
};
export const fetchWrapper = (url, payload) => {
  payload = {
    ...defaultRequest,
    ...payload,
  };
  return fetch(url, payload).then(
    (response) => {
      // Check that the response is valid and reject an error
      // response to prevent subsequent attempt to parse json
      if (!response.ok) {
        if (response.status === 500) {
          message.error(
            intl.formatMessage({
              id: "nemo.error500msg",
            })
          );
        } else if (response.status === 503) {
          message.error(
            intl.formatMessage({
              id: "nemo.error503msg",
            })
          );
        } else if (response.status === 504) {
          message.error(
            intl.formatMessage({
              id: "nemo.error504msg",
            })
          );
        } else {
          message.error(
            intl.formatMessage({
              id: "nemo.somethingWentWrongPleaseTryAgain",
            })
          );
        }
        return;
      }
      if (response.status === 403) {
        history.push("/access-denied");
      }
      return response.json ? response.json() : response;
    },
    (error) => {
      /**
       * user url, payload and error here for sending to the logger.
       */
      if (error instanceof TypeError) {
        window.location.href = baseURL.API_URL_BASE + "redirect";
      }
      console.error("fetch error : ", error);
    }
  );
};
