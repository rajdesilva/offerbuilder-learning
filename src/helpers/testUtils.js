import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render as rtlRender } from "@testing-library/react";
import { getValue } from "./utility";
import rootReducers from "../redux/rootReducers";
import * as languages from "../langs";
import { IntlProvider } from "react-intl";
const initialReducerState = {};
window.getValue = getValue;

window.google = {
  maps: {
    Marker: class {},
    Map: class {},
    LatLngBounds: class {},
    places: {
      Autocomplete: class {},
      AutocompleteService: class {},
      PlacesServiceStatus: {
        INVALID_REQUEST: "INVALID_REQUEST",
        NOT_FOUND: "NOT_FOUND",
        OK: "OK",
        OVER_QUERY_LIMIT: "OVER_QUERY_LIMIT",
        REQUEST_DENIED: "REQUEST_DENIED",
        UNKNOWN_ERROR: "UNKNOWN_ERROR",
        ZERO_RESULTS: "ZERO_RESULTS",
      },
      PlacesAutocomplete: {
        INVALID_REQUEST: "INVALID_REQUEST",
        NOT_FOUND: "NOT_FOUND",
        OK: "OK",
        OVER_QUERY_LIMIT: "OVER_QUERY_LIMIT",
        REQUEST_DENIED: "REQUEST_DENIED",
        UNKNOWN_ERROR: "UNKNOWN_ERROR",
        ZERO_RESULTS: "ZERO_RESULTS",
      },
    },

    MarkerClusterer: class {},
    Geocoder: class {},
  },
};
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

function render(
  ui,
  {
    initialState = initialReducerState,
    store = createStore(rootReducers, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <IntlProvider locale="en" messages={languages["en"]}>
          {children}
        </IntlProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
