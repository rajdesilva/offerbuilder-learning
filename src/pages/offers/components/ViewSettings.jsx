import React, { Fragment, useEffect } from "react";
import { Row, Col, Divider, Skeleton } from "antd";
import { useSelector } from "react-redux";
import moment from "moment/min/moment-with-locales";
import { FormattedMessage } from "react-intl";
import ViewDeepLinkSettings from "./ViewDeepLinkSettings";
import ViewTarget from "./ViewTarget";
import { SettingsPropertyList } from "./SettingsPropertyList";
import { isEqual } from "lodash";
import { useRouteMatch } from "react-router";
import { getOfferDetails } from "../service";
import { checkIfUserHasRole } from "../../../helpers/utility";
import { appConstants } from "../../../common";

function ViewSettings() {
  const match = useRouteMatch({
    path: "/offers/view/:offerId?",
    strict: true,
    sensitive: true,
  });
  const loading = useSelector(
    (state) =>
      window.getValue(state, "offerlistsearchandfilters.offerDetailsLoading"),
    isEqual
  );
  const newoffersettingsparam = useSelector(
    (state) => window.getValue(state, "newoffersettingsparam"),
    isEqual
  );
  useEffect(() => {
    if (match) {
      getOfferDetails(window.getValue(match, "params.offerId"));
    }
  }, []);
  return (
    <Skeleton loading={loading} active>
      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.internamName" />
        </Col>
        <Col span={8}>
          <div
            className="offerid-view-field-name"
            data-testid="offer-id-display-field"
          >
            {window.getValue(newoffersettingsparam, "name")}
          </div>
        </Col>
      </Row>

      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.status" />
        </Col>
        <Col span={4}>
          <div
            className="offerid-view-field-name"
            data-testid="offer-id-display-field"
          >
            {window.getValue(newoffersettingsparam, "status.name")}
          </div>
        </Col>
      </Row>

      {checkIfUserHasRole(appConstants.USER_ROLE.ADMIN) ? (
        <Row gutter={[8, 8]}>
          <Col span={4} className="offerid-view-field">
            <FormattedMessage id="nemo.type" />
          </Col>

          <Col span={14}>
            <div className="offerid-view-field-name" data-testid='offer-type'>
              {window.getValue(newoffersettingsparam, "type")}
            </div>
          </Col>
        </Row>
      ) : (
        <Fragment />
      )}

      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.offerID" />
        </Col>

        <Col span={6}>
          <div
            className="offerid-view-field-name"
            data-testid="offer-id-display-field"
          >
            {window.getValue(newoffersettingsparam, "offerId")}
          </div>
        </Col>
      </Row>

      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.bookingDates" />
        </Col>
        <Col span={5}>
          <div
            className="offerid-view-field-name"
            data-testid="offer-id-display-field"
          >
            {window.getValue(
              newoffersettingsparam,
              "bookingDateRange.startDate"
            ) ? (
              <Fragment>
                {moment(
                  window.getValue(
                    newoffersettingsparam,
                    "bookingDateRange.startDate"
                  )
                ).format("L")}
                --
                {moment(
                  window.getValue(
                    newoffersettingsparam,
                    "bookingDateRange.endDate"
                  )
                ).format("L")}{" "}
              </Fragment>
            ) : (
              "--"
            )}
          </div>
        </Col>
        <Col
          span={
            window.getValue(newoffersettingsparam, "bookingDatesTimezone.value")
              ? 6
              : 0
          }
        >
          <div
            className="offerid-view-field-name"
            data-testid="offer-id-display-field"
          >
            -- &nbsp; &nbsp;
            {window.getValue(
              newoffersettingsparam,
              "bookingDatesTimezone.value"
            ) || "--"}
          </div>
        </Col>
      </Row>

      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.travellingDates" />
        </Col>
        <Col span={6}>
          <div
            className="offerid-view-field-name"
            data-testid="offer-id-display-field"
          >
            {window.getValue(
              newoffersettingsparam,
              "travellingDateRange.startDate"
            ) ? (
              <Fragment>
                {moment(
                  window.getValue(
                    newoffersettingsparam,
                    "travellingDateRange.startDate"
                  )
                ).format("L")}
                --
                {moment(
                  window.getValue(
                    newoffersettingsparam,
                    "travellingDateRange.endDate"
                  )
                ).format("L")}{" "}
              </Fragment>
            ) : (
              "--"
            )}
          </div>
        </Col>
      </Row>

      <Divider />
      <Row gutter={[8, 8]}>
        <Col span={5}>
          <h3>
            <FormattedMessage id="nemo.travelProduct" />
          </h3>
        </Col>
      </Row>
      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.properties" />
        </Col>
        <Col span={20}>
          <SettingsPropertyList isForViewOnly={true} />
        </Col>
      </Row>
      <Divider />
      <Row gutter={[8, 8]}>
        <Col span={4} className="offerid-view-field">
          <FormattedMessage id="nemo.target" />
        </Col>
        <Col span={20}>
          <ViewTarget />
        </Col>
      </Row>
      <Divider />
      <ViewDeepLinkSettings />
    </Skeleton>
  );
}

export default ViewSettings;
