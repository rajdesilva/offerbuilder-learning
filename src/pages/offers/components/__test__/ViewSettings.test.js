import React from "react";
import { createStore } from "redux";
import { render } from "../../../../helpers/testUtils";
import { MemoryRouter } from "react-router-dom";
import * as Util from "../../../../helpers/utility/checkIfUserHasRole";
import ViewSettings from "../ViewSettings";
import { tempStore } from "./tempStore";

jest.mock("./../../service");

describe.only("ViewSettings component test", () => {
  test("create ViewSettings components snapshot", () => {
    const store = createStore(() => ({ ...tempStore }));
    const wrapper = render(
      <MemoryRouter initialEntries={["/offers/view/123"]}>
        <ViewSettings />
      </MemoryRouter>,
      {
        store,
      }
    );
    const { container } = wrapper;
    expect(container.firstChild).toMatchSnapshot();
  });

  test("ViewSettings, incase of admin, offer type is displayed", () => {
    const store = createStore(() => ({ ...tempStore }));
    const checkIfUserHasRole = jest.spyOn(Util, "checkIfUserHasRole");
    checkIfUserHasRole.mockReturnValue(true);
    const { getByTestId } = render(
      <MemoryRouter initialEntries={["/offers/view/123"]}>
        <ViewSettings />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(getByTestId('offer-type')).toBeInTheDocument()
  });

  test("ViewSettings, incase of users other than admin, offer type is hidden", () => {
    const store = createStore(() => ({ ...tempStore }));
    const checkIfUserHasRole = jest.spyOn(Util, "checkIfUserHasRole");
    checkIfUserHasRole.mockReturnValue(false);
    const { queryByTestId } = render(
      <MemoryRouter initialEntries={["/offers/view/123"]}>
        <ViewSettings />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(queryByTestId('offer-type')).not.toBeInTheDocument()
  });
});
