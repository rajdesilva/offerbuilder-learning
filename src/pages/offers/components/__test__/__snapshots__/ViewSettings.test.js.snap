// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSettings component test create ViewSettings components snapshot 1`] = `
<div
  class="ant-row"
  style="margin: -4px -4px 4px -4px;"
>
  <div
    class="ant-col ant-col-4 offerid-view-field"
    style="padding: 4px 4px 4px 4px;"
  >
    Internal Name
  </div>
  <div
    class="ant-col ant-col-8"
    style="padding: 4px 4px 4px 4px;"
  >
    <div
      class="offerid-view-field-name"
      data-testid="offer-id-display-field"
    >
      Cypress Offer - Edit
    </div>
  </div>
</div>
`;
